File: core/r/tests/testthat/test-correlation-heatmap.R
# The correlation heatmap recipe: a base-graphics draw closure plus a pairwise
# stats table with one row per variable pair.
test_that("recipe_correlation_heatmap builds a pairwise stats table", {
  set.seed(2)
  f <- rnorm(60)
  df <- data.frame(
    a = f + rnorm(60, sd = 0.2),
    b = f + rnorm(60, sd = 0.2),
    c = -f + rnorm(60, sd = 0.2),
    d = rnorm(60)
  )
  opt <- list(recipe = "correlation_heatmap", data = "vars.csv", test = "pearson")
  spec <- spec_from_opt(opt)

  res <- recipe_correlation_heatmap(df, spec)

  expect_true(is.function(res$draw))
  expect_null(res$plot)
  expect_equal(nrow(res$stats), 6)   # 4 variables -> 6 pairs
  expect_true(max(res$stats$r) > 0.8)
  expect_true(min(res$stats$r) < -0.8)
  expect_equal(res$test_meta$n_variables, 4)
})

File: core/python/tests/test_correlation_heatmap.py
from types import SimpleNamespace

import numpy as np
import pandas as pd

from pubplot.recipes.correlation_heatmap import recipe_correlation_heatmap
from pubplot.spec import spec_from_args


def test_recipe_correlation_heatmap_end_to_end():
    rng = np.random.default_rng(1)
    f = rng.normal(0, 1, 60)
    df = pd.DataFrame({
        "a": f + rng.normal(0, 0.2, 60),
        "b": f + rng.normal(0, 0.2, 60),
        "c": -f + rng.normal(0, 0.2, 60),
        "d": rng.normal(0, 1, 60),
    })
    a = SimpleNamespace(recipe="correlation_heatmap", data="vars.csv",
                        x=None, y=None, test="pearson")
    spec = spec_from_args(a)

    res = recipe_correlation_heatmap(df, spec)

    # 4 variables -> 6 unique pairs.
    assert len(res["stats"]) == 6
    assert set(res["stats"].columns) >= {"var1", "var2", "r", "p_value", "p_adj", "significant"}
    assert res["stats"]["r"].max() > 0.8    # a & b co-vary
    assert res["stats"]["r"].min() < -0.8   # c is anti-correlated
    assert res["test_meta"]["n_variables"] == 4
